Add gometalinter to travis ci #20

u5surf opened the pull request:
#20 
I added in gometalinter to travisci in this repo.
Gometalinter uses the following lint method.
```
gometalinter --concurrency=4 --enable-gc --deadline=300s --disable-all 
--enable=deadcode
--enable=errcheck
--enable=goconst
--enable=gofmt 
--enable=goimports
--enable=golint
--exclude=.pb.go
--enable=gotype
--enable=ineffassign
--enable=interfacer
--enable=lll --line-length=120 
--enable=misspell 
--enable=structcheck 
--enable=unconvert 
--enable=unused 
--enable=varcheck 
--enable=vet 
--enable=maligned 
--vendor ./...
```
https://github.com/alecthomas/gometalinter#supported-linters

I would like to know whether you want to use supported linters else.

nitishm commented:
Thanks for this PR @u5surf. 
My thought looking at this is to move `go vet ` and `staticcheck` into `gometalinter` and also instead of passing the `enable` list, we make use of `.gometalinter.json` config file.

Shivam010 commented:
I think [golangci-lint](https://github.com/golangci/golangci-lint) will serve much better than gometalinter.

Afterall,on an average, golangci-lint is 4.6 times faster than gometalinter.

> ### Comparison with gometalinter
> 
> We compare golangci-lint and gometalinter in default mode, but explicitly enable all linters because of small differences in the default configuration.
> 
> ```bash
> $ golangci-lint run --no-config --issues-exit-code=0 --deadline=30m \
>   --disable-all --enable=deadcode  --enable=gocyclo --enable=golint --enable=varcheck \
>   --enable=structcheck --enable=maligned --enable=errcheck --enable=dupl --enable=ineffassign \
>   --enable=interfacer --enable=unconvert --enable=goconst --enable=gosec --enable=megacheck
> $ gometalinter --deadline=30m --vendor --cyclo-over=30 --dupl-threshold=150 \
>   --exclude=<default golangci-lint excludes> --skip=testdata --skip=builtin \
>   --disable-all --enable=deadcode  --enable=gocyclo --enable=golint --enable=varcheck \
>   --enable=structcheck --enable=maligned --enable=errcheck --enable=dupl --enable=ineffassign \
>   --enable=interfacer --enable=unconvert --enable=goconst --enable=gosec --enable=megacheck
>   ./...
> ```
> 
> | Repository | GolangCI Time | GolangCI Is Faster than Gometalinter | GolangCI Memory | GolangCI eats less memory than Gometalinter |
> | ---------- | ------------- | ------------------------------------ | --------------- | ------------------------------------------- |
> | gometalinter repo, 4 kLoC   | 6s    | **6.4x** | 0.7GB | 33%  |
> | self-repo, 4 kLoC           | 12s   | **7.5x** | 1.2GB | 41%  |
> | beego, 50 kLoC              | 10s   | **4.2x** | 1.4GB | 9%   |
> | hugo, 70 kLoC               | 15s   | **6.1x** | 1.6GB | 44%  |
> | consul, 127 kLoC            | 58s   | **4x**   | 2.7GB | 41%  |
> | terraform, 190 kLoC         | 2m13s | **1.6x** | 4.8GB | 0%   |
> | go-ethereum, 250 kLoC       | 33s   | **5x**   | 3.6GB | 0%   |
> | go source (`$GOROOT/src`), 1300 kLoC | 2m45s | **2x** | 4.7GB | 0% |
> 
> **On average golangci-lint is 4.6 times faster** than gometalinter. Maximum difference is in the
> self-repo: **7.5 times faster**, minimum difference is in terraform source code repo: 1.8 times faster.
> 
> On average golangci-lint consumes 26% less memory.

Also, default linters of golangci-lint are sufficiently good for the repo.

> ### Enabled By Default Linters
> 
> - [govet](https://golang.org/cmd/vet/) - Vet examines Go source code and reports suspicious constructs, such as Printf calls whose arguments do not align with the format string
> - [errcheck](https://github.com/kisielk/errcheck) - Errcheck is a program for checking for unchecked errors in go programs. These unchecked errors can be critical bugs in some cases
> - [staticcheck](https://staticcheck.io/) - Staticcheck is a go vet on steroids, applying a ton of static analysis checks
> - [unused](https://github.com/dominikh/go-tools/tree/master/cmd/unused) - Checks Go code for unused constants, variables, functions and types
> - [gosimple](https://github.com/dominikh/go-tools/tree/master/cmd/gosimple) - Linter for Go source code that specializes in simplifying a code
> - [structcheck](https://github.com/opennota/check) - Finds an unused struct fields
> - [varcheck](https://github.com/opennota/check) - Finds unused global variables and constants
> - [ineffassign](https://github.com/gordonklaus/ineffassign) - Detects when assignments to existing variables are not used
> - [deadcode](https://github.com/remyoudompheng/go-misc/tree/master/deadcode) - Finds unused code
> - typecheck - Like the front-end of a Go compiler, parses and type-checks Go code
> 
> ### Disabled By Default Linters (`-E/--enable`)
> 
> - [golint](https://github.com/golang/lint) - Golint differs from gofmt. Gofmt reformats Go source code, whereas golint prints out style mistakes
> - [gosec](https://github.com/securego/gosec) - Inspects source code for security problems
> - [interfacer](https://github.com/mvdan/interfacer) - Linter that suggests narrower interface types
> - [unconvert](https://github.com/mdempsky/unconvert) - Remove unnecessary type conversions
> - [dupl](https://github.com/mibk/dupl) - Tool for code clone detection
> - [goconst](https://github.com/jgautheron/goconst) - Finds repeated strings that could be replaced by a constant
> - [gocyclo](https://github.com/alecthomas/gocyclo) - Computes and checks the cyclomatic complexity of functions
> - [gofmt](https://golang.org/cmd/gofmt/) - Gofmt checks whether code was gofmt-ed. By default this tool runs with -s option to check for code simplification
> - [goimports](https://godoc.org/golang.org/x/tools/cmd/goimports) - Goimports does everything that gofmt does. Additionally it checks unused imports
> - [maligned](https://github.com/mdempsky/maligned) - Tool to detect Go structs that would take less memory if their fields were sorted
> - [megacheck](https://github.com/dominikh/go-tools/tree/master/cmd/megacheck) - 3 sub-linters in one: unused, gosimple and staticcheck
> - [depguard](https://github.com/OpenPeeDeeP/depguard) - Go linter that checks if package imports are in a list of acceptable packages
> - [misspell](https://github.com/client9/misspell) - Finds commonly misspelled English words in comments
> - [lll](https://github.com/walle/lll) - Reports long lines
> - [unparam](https://github.com/mvdan/unparam) - Reports unused function parameters
> - [nakedret](https://github.com/alexkohler/nakedret) - Finds naked returns in functions greater than a specified function length
> - [prealloc](https://github.com/alexkohler/prealloc) - Finds slice declarations that could potentially be preallocated
> - [scopelint](https://github.com/kyoh86/scopelint) - Scopelint checks for unpinned variables in go programs
> - [gocritic](https://github.com/go-critic/go-critic) - The most opinionated Go source code linter
> - [gochecknoinits](https://github.com/leighmcculloch/gochecknoinits) - Checks that no init functions are present in Go code
> - [gochecknoglobals](https://github.com/leighmcculloch/gochecknoglobals) - Checks that no globals are present in Go code

*Quotes taken from https://github.com/golangci/golangci-lint/blob/master/README.md

u5surf commented:
> Thanks for this PR @u5surf.
> My thought looking at this is to move `go vet ` and `staticcheck` into `gometalinter` and also instead of passing the `enable` list, we make use of `.gometalinter.json` config file.

@Shivam010  said golangci-lint is faster than gometalinter. so, I changed golangci-lint
and I move `go vet` and  `staticcheck`  into  `golangci-lint`. Moreover I added `.golangci.yml` config file which can set options flexiblily.
e.g. https://github.com/golangci/golangci-lint#config-file

nitishm commented on .golangci.yml:
```diff
@@ -0,0 +1,33 @@
+    check-shadowing: true
+  golint:
+    min-confidence: 0
+  gocyclo:
```
If we are disabling `gocyclo` in L#25 this is not needed.

nitishm commented on .golangci.yml:
```diff
@@ -0,0 +1,33 @@
+
+linters:
+  enable-all: true
+  disable:
```
Did you disable these based on the complexity of resolving errors with all enabled by default ? I can imagine that the example file might be causing some severe issues.

nitishm commented on examples/json_obj/json_obj.go:
```diff
@@ -78,8 +82,8 @@ func main() {
 	res, err = rejson.JSONGet(conn, "obj", ".",
-		&rejson.JSONGetOptionIndent{"\t"}, &rejson.JSONGetOptionNewLine{"\n"},
-		&rejson.JSONGetOptionSpace{" "}, &rejson.JSONGetOptionNoEscape{})
+		rejson.NewJSONGetOptionIndent("\t"), rejson.NewJSONGetOptionNewLine("\n"),
```
Also you might want to pull first before submitting this as I am merging https://github.com/nitishm/go-rejson/pull/22 before I merge these changes.

u5surf commented:
@nitishm 
I added several linter.
- prealloc
- gosec
   it's ok by default   
- scopelint
   fix range scope in function literal in rejson_test.go
https://github.com/nitishm/go-rejson/blob/d020e35baee5a8d4715869366d411fcbb3562763/rejson_test.go#L144-L156
I know it is better the using argument of  the anonymous function than substituting another variable.
but tt doesn't declare the struct types such as` type TestCase struct{}`.
- gocritic
   change if-else to switch statement in rejson.go
https://github.com/nitishm/go-rejson/blob/d020e35baee5a8d4715869366d411fcbb3562763/rejson.go#L402-L409

and fix rule of minum cyclo complexity 10 to 15 
because, it is failed only in rejson_test.go
generally, It can be still refactor rejson_test.go as an another issue to success depl, more low  cyclo complexity.

nitishm commented:
Not a fan of scopelint. Let's revert `rejson_test.go` and disable `scopelint` in the config file. Everything else looks good.

nitishm commented:
Awesome looks great !
